### PR TITLE
Better ModelDefaults values.

### PR DESCRIPTION
**Pull request by tlm**
This change refactors our new ModelDefaultValues to not be hierarchical internal and instead rely on the defaults service to do the building of the hierarchy.  When first investigating this work I made the mistake of reading the current defaults code in state and seeing that we only had 3 sources of defaults defined in Juju and model the problem as such.

After starting to move the model creation process over to the new service layer we actually have a few more sources of defaults that have appeared in Juju over the years that did not get built into the standards setup or refactored.  See [here](https://github.com/juju/juju/blob/03ceaf7172e08b9341046b02c0d378ab415f4920/apiserver/facades/client/modelmanager/modelmanager.go#L221) for example.

To deal with this problem and so that we can carrel all of these opinions down into one service to keep them under control the defaults value has been updated to just represent one value of what model defaults thinks the value should be when all is said and done.

This change also allows the provider of the default value to offer up logic to model config on how the default should be applied when the value is set in both places. This is going to be useful in a follow up PR.

For the moment this is a mechanical change to support the next lot of changes coming through that gets the other defaults sources out of the API Server.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Unit tests in domain/modeldefaults/service and domain/modelconfig/service

These domain services are not integrated at the moment so there is no manual testing that can be performed. Stay tuned.

## Documentation changes

N/A

## Links

**Jira card:** JUJU-4647

### CODE REVIEW

File: domain/modeldefaults/bootstrap/bootstrap_test.go [hunk 1]
```diff
@@ -34,20 +32,22 @@ func (_ *bootstrapSuite) TestBootstrapModelDefaults(c *gc.C) {
...
-		"default": modeldefaults.DefaultAttributeValue{
-			Default: "some value",
-		},
-		"controller": modeldefaults.DefaultAttributeValue{
-			Controller: "some value",
-		},
-		"region": modeldefaults.DefaultAttributeValue{
-			Region: "some value",
-		},
-	})
```
Comment (SimonRichardson) [inline domain/modeldefaults/bootstrap/bootstrap_test.go:52]: Fix me.

File: domain/modeldefaults/types.go [hunk 2]
```diff
@@ -5,34 +5,81 @@ package modeldefaults
...
-	Default any
+	// Strategy is the ApplyStrategy that should be used when deciding how to
+	// integrate this default value. If Strategy is the zero value then consult
+	// [DefaultAttributeValue.ApplyStrategy] for expected behaviour.
+	Strategy ApplyStrategy
 
-	// Region represents a value that come from a model's set cloud region.
-	Region any
+	// V is the default value.
+	V any
```
Comment (SimonRichardson) [inline domain/modeldefaults/types.go:38]: Not sure I like the name `V`...

**Review by SimonRichardson — COMMENTED** [review]
I like the way this is going. Would be good to get @wallyworld to look at this, as he's got the history for this.

File: domain/modeldefaults/types.go [hunk 3]
```diff
@@ -5,34 +5,81 @@ package modeldefaults
...
+// config and the ultimate say in what the value is DefaultAttributeValue
+// provides the mechanism for sources to place their opinions in one place and
+// for the consuming side (model config) to have a one size fits all approach
+// for taking on the defaults.
 type DefaultAttributeValue struct {
-	// Controller represents a value that comes from the controller,
-	// specifically the cloud.
-	Controller any
+	// Source describes the source of the default value.
+	Source string
```
[the remark this answers is not in the document]
Comment (tlm) [reply domain/modeldefaults/types.go:30]: Not convinced yet but I have it in my mind should I see more justification. At the moment it's a display problem that has leached it's way into the problem domain.

File: domain/modeldefaults/types_test.go [hunk 4]
```diff
@@ -12,73 +12,120 @@ type typesSuite struct{}
...
+func (t *testApplyStrategy) Apply(d, s any) any {
+	t.Called = true
+	return s
+}
 
-	has, source = val.Has("default")
-	c.Assert(has, jc.IsFalse)
-	c.Assert(source, gc.Equals, "")
+// TestApplyStrategy is checking to make sure that if we set an apply strategy
+// on the DefaultAttributeValue. This test isn't concerned about testing the
```
Comment (wallyworld) [inline domain/modeldefaults/types_test.go:118]: sentence doesn't parse

File: domain/modeldefaults/types_test.go [hunk 5]
```diff
@@ -12,73 +12,120 @@ type typesSuite struct{}
...
+	has, source = val.Has(&struct{ name string }{"test"})
+	c.Check(has, jc.IsFalse)
+	c.Check(source, gc.Equals, "")
+}
+
+// testApplyStrategy is a test implementation of ApplyStrategy that is here to
+// just indicate that the Apply method of the strategy has been called.
+type testApplyStrategy struct {
+	// Called indicates that the Apply func of this struct has been called.
+	Called bool
```
Comment (wallyworld) [inline domain/modeldefaults/types_test.go:108]: why export?

**Review by wallyworld — APPROVED** [review]

**Pull request merged**